File: 216-combination-sum-iii/combination-sum-iii.py
class Solution:
    def combinationSum3(self, k: int, n: int) -> List[List[int]]:
        def solve(i,k,n,curr,ans,total):
            if(len(curr) == k):
                if(total == n):
                    ans.append(curr.copy())
                return
            if i > 9:
                return
            curr.append(i)
            solve(i+1,k,n,curr,ans,total+i)
            curr.pop()
            solve(i+1,k,n,curr,ans,total)
        curr = []
        ans = []
        solve(1,k,n,curr,ans,0)
        return ans

File: 216-combination-sum-iii/combination-sum-iii.cpp
class Solution {
public:
    vector<vector<int>> combinationSum3(int k, int n) {
        vector<int>curr;
        vector<vector<int>>ans;
        solve(1,k,n,curr,ans,0);
        return ans;
    }
    void solve(int i, int k, int n, vector<int>&curr, vector<vector<int>>&ans, int sum){
        if(curr.size() == k){
            if(sum == n){
                ans.push_back(curr);
            }
            return;
        }
        if(i>9) return;
        curr.push_back(i);
        solve(i+1,k,n,curr,ans,sum+i);
        curr.pop_back();
        solve(i+1,k,n,curr,ans,sum);
    }
};